packages/api/src/command/medical/cq-directory/rebuild-cq-directory.ts

```diff
@@ -0,0 +1,105 @@
+  host: dbCreds.host,
+  port: dbCreds.port,
+  dialect: dbCreds.engine,
+});
+
+export const rebuildCQDirectory = async (failGracefully = false): Promise<void> => {
+  let currentPosition = 0;
+  let isDone = false;
+
+  try {
+    await createTempCQDirectoryTable();
+    while (!isDone) {
+      try {
+        const apiKey = Config.getCQApiKey();
+        const cq = new Carequality(apiKey);
+        const orgs = await cq.listOrganizations({ start: currentPosition });
+        if (orgs.length < BATCH_SIZE) isDone = true; // if CQ directory returns less than BATCH_SIZE number of orgs, that means we've hit the end
+        currentPosition += BATCH_SIZE;
+        const parsedOrgs = parseCQDirectoryEntries(orgs);
+        await bulkInsertCQDirectoryEntries(sequelize, parsedOrgs);
+        console.log(`Added ${currentPosition} CQ directory entries...`);
+        await sleep(750);
+      } catch (error) {
+        isDone = true;
+        if (!failGracefully) {
+          throw error;
+        }
+      }
+    }
+  } catch (error) {
+    await deleteTempCQDirectoryTable();
+    const msg = `Failed to rebuild the directory`;
+    console.log(`${msg}, error: ${error}`);
+    capture.message(msg, {
+      extra: { context: `rebuildCQDirectory`, error },
+      level: "error",
+    });
+    throw error;
+  }
+  try {
+    await renameCQDirectoryTablesAndUpdateIndexes();
+  } catch (error) {
+    const msg = `Failed the last step of CQ directory rebuild`;
+    console.log(`${msg}. Cause: ${error}`);
+    capture.message(msg, {
+      extra: { context: `renameCQDirectoryTablesAndUpdateIndexes`, error },
+      level: "error",
+    });
+    throw error;
+  }
+};
+
+const createTempCQDirectoryTable = async () => {
+  const query = `CREATE TABLE IF NOT EXISTS ${cqDirectoryEntryTemp} AS SELECT * FROM ${cqDirectoryEntry} WHERE 1=0;`;
+  await sequelize.query(query, {
+    type: QueryTypes.INSERT,
+  });
+};
+
+const deleteTempCQDirectoryTable = async () => {
+  const query = `DROP TABLE IF EXISTS ${cqDirectoryEntryTemp}`;
+  await sequelize.query(query, {
+    type: QueryTypes.DELETE,
+  });
+};
+
+const renameCQDirectoryTablesAndUpdateIndexes = async () => {
+  const dropBackupQuery = `DROP TABLE IF EXISTS ${cqDirectoryEntryBackup};`;
+  await sequelize.query(dropBackupQuery, {
+    type: QueryTypes.DELETE,
+    logging: false,
+  });
+
+  const renameTablesQuery = `
+    ALTER TABLE ${cqDirectoryEntry} RENAME TO ${cqDirectoryEntryBackup};
+    ALTER TABLE ${cqDirectoryEntryTemp} RENAME TO ${cqDirectoryEntry};
+    ALTER INDEX IF EXISTS cq_directory_entry_pkey RENAME TO cq_directory_entry_backup_pkey;`;
+  await sequelize.query(renameTablesQuery, {
+    type: QueryTypes.UPDATE,
+    logging: false,
+  });
+
+  const createIndexQuery = `CREATE INDEX IF NOT EXISTS cq_directory_entry_pkey ON ${cqDirectoryEntry} (id);`;
+  await sequelize.query(createIndexQuery, {
+    type: QueryTypes.INSERT,
+    logging: false,
+  });
```

Comment:
💯 transactions, thx!

Reply:
I don't know about the locks in this case. 

Thinking about the risks, we don't care too much about someone grabbing "old data" in the moment of renaming. This is unlikely, and isn't very critical.

Additionally, the renaming process of both tables in one transaction is an extremely quick process, since it's only a metadata update.

Lastly, I haven't been able to find READ locks on PostgreSQL. 

Thoughts?

Reply:
I suggest we use `executeOnDBTx()` it has a battle tested version of the code. See how it needs to have a `finally` instruction.

As for the lock, I think we need a table lock with `exclusive` access type, so no other connection can read from the table while we're doing it - from [this page](https://hevodata.com/learn/postgresql-locks/#table):

```sql
Lock table ${cqDirectoryEntry} ACCESS EXCLUSIVE;
```

It's simple to test this:
- build a test endpoint to query some records from the table cqDirectoryEntry
- add a "sleep" after the execution of the `renameTablesQuery`
- execute the `renameCQDirectoryTablesAndUpdateIndexes` function and, while it sleeps, execute the endpoint to query records from the table (it should be blocked and only finish executing when the renaming is done.